### PR TITLE
Zone orientation effect on keylight, skybox and ambient light.

### PR DESCRIPTION
**Pull request by NissimHadar**
# Description
This PR implements rotates the keylight, skybox and ambient light by the zone orientation.
# Testing
1.  After installing the PR, verify that the local domain appears as before (to verify usage with older content).  
2.  Download the GitHub test repository from https://github.com/NissimHadar/hifi_tests.git - create and then checkout  the **hazeTests** branch.
3.  Run setup.bat in the tests/tools/autoTester folder. This will download the autoTester executable and required resources into tests/tools/autoTester/Release.

4.  Move to an empty area in a domain with full permissions (such as localhost/5000,5000,5000).

5.  Run script: tests/content/entity/zone/zoneOrientation/runAuto.js

tests/content/entity/zone/zoneOrientationwill now contain 30 images.  15 of these are the expected results named ExpectedImage_00000.jpg through ExpectedImage_00014.jpg.
The results are the 15 images named `hifi-snap-by........jpg. The images should match up (1st == 1st and so on).

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5918/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5918/HighFidelity-Beta-PR12083-11cef7c2492c19068722f068697f8d9a25e40cd4.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5918/HighFidelity-Beta-PR12083-11cef7c2492c19068722f068697f8d9a25e40cd4.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/5919/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5919/HighFidelity-Beta-PR12083-a1b96fc5244f9b6c1134cf77e332c1e55fada061.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5919/HighFidelity-Beta-PR12083-a1b96fc5244f9b6c1134cf77e332c1e55fada061.dmg

File: libraries/model/src/model/Skybox.cpp [hunk 1]
```diff
@@ -37,6 +39,12 @@ void Skybox::setCubemap(const gpu::TexturePointer& cubemap) {
     }
 }
 
+void Skybox::setOrientation(glm::quat rotation) {
+    // The zone rotations need to be negated
```
Comment (AndrewMeadows) [inline libraries/model/src/model/Skybox.cpp]: Negating the real component of the quaternion provides its inverse.  So the comment should be:

`// the zone rotation needs to be the inverse`

However, maybe it should be explicit that the rotation being stored in the Skybox is an inverse:

`Skybox::_rotation` --> `Skybox::_invRotation`

File: libraries/entities-renderer/src/RenderableZoneEntityItem.cpp [hunk 2]
```diff
@@ -319,7 +319,7 @@ void ZoneEntityRenderer::updateKeySunFromEntity() {
     // Set the keylight
     sunLight->setColor(ColorUtils::toVec3(_keyLightProperties.getColor()));
     sunLight->setIntensity(_keyLightProperties.getIntensity());
-    sunLight->setDirection(_keyLightProperties.getDirection());
+    sunLight->setDirection(entity->getTransform().getRotation() * _keyLightProperties.getDirection());
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderableZoneEntityItem.cpp]: This method takes an entity but only uses it to extract a rotation.  Maybe this should be `updateKeySun()` and take a `rotation` as argument?

File: libraries/model/src/model/Skybox.h [hunk 3]
```diff
@@ -61,6 +63,8 @@ class Skybox {
 
     mutable gpu::BufferView _schemaBuffer;
     gpu::TexturePointer _cubemap;
+    static glm::quat _rotation;
```
Comment (AndrewMeadows) [inline libraries/model/src/model/Skybox.h]: It isn't clear to me why `Skybox::_rotation` is `static`.  There is only one _rotation for all Skybox instances?  Maybe its name should be something that makes it clear that it is a global setting?  `_globalInvRotation`?

If `Skybox::_rotation` really should be `static` then I think `Skybox::setOrientation()` should also be `static`, and maybe its name should be changed to something like: `Skybox::setGlobalInvOrientation()` or... something that makes it more clear what it actually does.

File: libraries/entities-renderer/src/RenderableZoneEntityItem.cpp (hunk 2 above)
Comment (NissimHadar) [reply libraries/entities-renderer/src/RenderableZoneEntityItem.cpp]: I agree.  It's funny that the old method had that name, and didn't have _any_ parameters...

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6816/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6816/HighFidelity-Beta-PR12083-1a9babab5d42c12ac95c0b000112ce906039cfc2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6816/HighFidelity-Beta-PR12083-1a9babab5d42c12ac95c0b000112ce906039cfc2.exe

**Comment by NissimHadar** [comment]
The render method is static - see line 48 in Skybox.h.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6852/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6852/HighFidelity-Beta-PR12083-1b1b93ec2a9b1d5d0ccfa0b583fd0b0e249af3da.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6882/HighFidelity-Beta-PR12083-19ecc5778c12ff45bb510014ef91a1cc37e547af.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6882/HighFidelity-Beta-PR12083-19ecc5778c12ff45bb510014ef91a1cc37e547af.dmg

File: libraries/graphics/src/graphics/Skybox.h [hunk 4]
```diff
@@ -61,6 +63,8 @@ class Skybox {
 
     mutable gpu::BufferView _schemaBuffer;
     gpu::TexturePointer _cubemap;
+    static glm::quat _rotation;
```
[the remark this answers is not in the document]
Comment (NissimHadar) [reply libraries/graphics/src/graphics/Skybox.h]: The method itself (render) is static.

File: libraries/graphics/src/graphics/Light.cpp [hunk 5]
```diff
@@ -158,3 +158,9 @@ void Light::setAmbientMapNumMips(uint16_t numMips) {
     _ambientSchemaBuffer.edit().mapNumMips = (float)numMips;
 }
 
+void Light::setZoneTransform(const glm::mat4& zoneTransform) {
```
Comment (samcake) [inline libraries/graphics/src/graphics/Light.cpp]: The light already has a transform (through setPosition setOrientation)
Why do you need another one just for the case of ambient ?
couldn;t we use the transform already there instead ?

It doesn;t need necessary in mey opinion to have a different member function for this?

File: libraries/graphics/src/graphics/Haze.cpp [hunk 6]
```diff
@@ -185,7 +185,7 @@ void Haze::setHazeBackgroundBlend(const float hazeBackgroundBlend) {
 void Haze::setZoneTransform(const glm::mat4& zoneTransform) {
     auto& params = _hazeParametersBuffer.get<Parameters>();
 
-    if (params.zoneTransform == zoneTransform) {
+    if (params.zoneTransform != zoneTransform) {
```
Comment (samcake) [inline libraries/graphics/src/graphics/Haze.cpp]: For haze, can you remove  'zone' from that member ?

File: libraries/graphics/src/graphics/Light.h [hunk 7]
```diff
@@ -149,6 +149,8 @@ class Light {
     void setAmbientMapNumMips(uint16_t numMips);
     uint16_t getAmbientMapNumMips() const { return (uint16_t) _ambientSchemaBuffer->mapNumMips; }
 
+    void setZoneTransform(const glm::mat4& zoneTransform);
```
Comment (samcake) [inline libraries/graphics/src/graphics/Light.h]: We don;t need a 'zone' specific  transform in the light
The light already has a transform (through setPosition setOrientation)
Why do you need another one just for the case of ambient ?
couldn;t we use the transform already there instead ?

File: libraries/graphics/src/graphics/Light.h [hunk 8]
```diff
@@ -162,7 +164,9 @@ class Light {
         float mapNumMips { 0.0f };
         float spare1;
         float spare2;
+
         gpu::SphericalHarmonics ambientSphere;
+        glm::mat4 zoneTransform;
```
Comment (samcake) [inline libraries/graphics/src/graphics/Light.h]: ok to have the member here, just don;t call it 'zone'

File: libraries/graphics/src/graphics/Light.slh [hunk 9]
```diff
@@ -34,7 +34,9 @@ vec3 getLightIrradiance(Light l) { return lightIrradiance_getIrradiance(l.irradi
 // Light Ambient
 struct LightAmbient {
     vec4 _ambient;
+
     SphericalHarmonics _ambientSphere;
+    mat4 zoneTransform;
```
Comment (samcake) [inline libraries/graphics/src/graphics/Light.slh]: same as above, don't call that transform 'zone'

File: libraries/graphics/src/graphics/Skybox.h [hunk 10]
```diff
@@ -37,6 +37,8 @@ class Skybox {
     void setCubemap(const gpu::TexturePointer& cubemap);
     const gpu::TexturePointer& getCubemap() const { return _cubemap; }
 
+    void setOrientation(glm::quat rotation);
```
Comment (jherico) [inline libraries/graphics/src/graphics/Skybox.h]: `const glm::quat&`

File: libraries/graphics/src/graphics/Skybox.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (jherico) [reply libraries/graphics/src/graphics/Skybox.h]: the static version takes as a parameter a `Skybox` instance.

File: libraries/graphics/src/graphics/Haze.cpp (hunk 6 above)
Comment (NissimHadar) [reply libraries/graphics/src/graphics/Haze.cpp]: Done.

File: libraries/graphics/src/graphics/Light.cpp (hunk 5 above)
Comment (NissimHadar) [reply libraries/graphics/src/graphics/Light.cpp]: I think we should keep it; I have renamed it `setTransform`.  It is different than setting the position or orientation.  The ambient light has no position or orientation, and having setTransform, to me, makes the intent clearer.

File: libraries/graphics/src/graphics/Light.h (hunk 7 above)
Comment (NissimHadar) [reply libraries/graphics/src/graphics/Light.h]: I have renamed to setTransform.  I think I don't understand why this is wrong.  We need the transform matrix as is.

File: libraries/graphics/src/graphics/Light.h (hunk 8 above)
Comment (NissimHadar) [reply libraries/graphics/src/graphics/Light.h]: Done.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6893/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6893/HighFidelity-Beta-PR12083-b59b81368cbc96758c72081cc3ba5d0ae9431d7e.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6893/HighFidelity-Beta-PR12083-b59b81368cbc96758c72081cc3ba5d0ae9431d7e.exe

File: libraries/graphics/src/graphics/Skybox.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (samcake) [reply libraries/graphics/src/graphics/Skybox.h]: this parameter can NOT be static.
it needs to  be a member parameter from the Skybox object that you access from the render function

File: libraries/graphics/src/graphics/Skybox.h [hunk 11]
```diff
@@ -37,6 +37,8 @@ class Skybox {
     void setCubemap(const gpu::TexturePointer& cubemap);
     const gpu::TexturePointer& getCubemap() const { return _cubemap; }
 
+    void setOrientation(const glm::quat& rotation);
```
Comment (samcake) [inline libraries/graphics/src/graphics/Skybox.h]: there should be a get function to the Orientation member of the skybox
so you can call it in the STATIC render method

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6909/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6909/HighFidelity-Beta-PR12083-2d80b8ad261f51eb8cc8c29edc3d2f0215652959.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6909/HighFidelity-Beta-PR12083-2d80b8ad261f51eb8cc8c29edc3d2f0215652959.exe

**Comment by NissimHadar** [comment]
I changed it differently: the variable should be mutable.  Adding a get
method would not help.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6921/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6921/HighFidelity-Beta-PR12083-24751a414ddbe39db765275ff71bb0ec014afa2c.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6921/HighFidelity-Beta-PR12083-24751a414ddbe39db765275ff71bb0ec014afa2c.exe

**Comment by sbennett77** [comment]
Currently Testing.

**Comment by sbennett77** [comment]
Passed.

@conklin94122

**Comment by samcake** [comment]
@marko8904 I can;t seem to be able to press merge
DO you know what's needed there ?
I see that the License/cla is not green yet

**Comment by marko8904** [comment]
@samcake - this commit https://github.com/highfidelity/hifi/pull/12083/commits/24751a414ddbe39db765275ff71bb0ec014afa2c and a few others seem to have Nissim as author but with an email that is not on his Github account. He'd just have to add that email to his Github. Otherwise CLA doesn't know that it is the same @NissimHadar . You can tell right away because the commits with unknown emails don't have any icons except the generic Octocat.

**Comment by NissimHadar** [comment]
Thanks.  I have added my second email to GitHub.

**Comment by marko8904** [comment]
@NissimHadar if you notice above there's a notice that the signature failed. Could you take a look at all the commits here without icons and see if there are any other emails that may be missing? I've just clicked on the "recheck" link but that didn't help meaning that something is still not right.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6993/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6993/HighFidelity-Beta-PR12083-1ac32ab444973fe6f7ee057a92cb4997e26c52fe.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6993/HighFidelity-Beta-PR12083-1ac32ab444973fe6f7ee057a92cb4997e26c52fe.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6994/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6995/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6995/HighFidelity-Beta-PR12083-d776593e926345690b55ed9d2cff0439a06cb60b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6995/HighFidelity-Beta-PR12083-d776593e926345690b55ed9d2cff0439a06cb60b.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/6996/app_12083_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/6996/HighFidelity-Beta-PR12083-d01f9cd1fbfa1b001484d946e566077de5945192.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/6996/HighFidelity-Beta-PR12083-d01f9cd1fbfa1b001484d946e566077de5945192.exe

**Pull request merged**